lib/elixir/lib/kernel.ex

```diff
@@ -957,7 +971,10 @@ defmodule Kernel do
       [1, 2 | 3]
 
   """
-  @spec (list ++ term) :: maybe_improper_list
+  @spec (nonempty_list ++ nonempty_list) :: nonempty_list
+  @spec (left ++ []) :: left when left: nonempty_list
+  @spec ([] ++ right) :: right when right: term
+  @spec (list ++ nonempty_improper_list(any, any)) :: nonempty_improper_list(any(), any())
```

Comment:
I think this spec got considerably harder to understand what it does. And it is a spec that does not matter because the function is inlined anyway, so it only exists for documentation purposes. So I prefer to keep the previous one.